src/main/scala/fs2/async/AsyncExt.scala

```diff
@@ -0,0 +1,208 @@
+package fs2.async
+
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicReference}
+
+import fs2.internal.Actor
+import fs2.util.Task.{Result, Callback}
+import fs2.{Strategy, Async}
+import fs2.util.{Task, Free}
+
+import scala.collection.immutable.Queue
+import scala.util.control.NonFatal
+
+
+/**
+ * Created by pach on 24/10/15.
+ */
+trait AsyncExt[F[_]] extends Async[F] {
+
+  /**
+   * Modifies `ref` by applying pure function `f` to create computation `F` that's value
+   * will be used to modify the `ref`, once computed.
+   *
+   * During evaluation of `F` all other modifications will be queued up,
+   * but the `get` will receive the `A` that was supplied to `f`.
+   *
+   * If the `ref` is not yet set, this queues up until `ref` is set for the first time.
+   *
+   * Note that order of queued modifications is not preserved
+   *
+   * Likewise with `set` Resulting task may complete before final value is applied to `ref`,
+   * but it is guaranteed that it will be applied before any other modify/set on the `ref`.
+   *
+   */
+  def modify[A](ref:Ref[A])(f: A => F[A]):F[(A,A)]
```

Comment:
These docs could use some cleanup and clarification. We can work on that later.

What is the pair that's returned? Is it `(old,new)`? Is `old` the argument that was passed to the `f`? Let's just introduce `case class Change[+A](previous: A, now: A)` and return a `Change[A]` here. You can put `Change` in the companion object of `AsyncExt`.